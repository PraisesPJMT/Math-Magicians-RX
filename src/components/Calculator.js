import React, { useState } from 'react';
import calculate from '../logic/calculate';
import Numbers from './cal_numbers';
import Operators from './cal_operators';
import './styles/Calculator.css';

const Calculator = () => {
  const [state, setState] = useState({ total: null, next: null, operation: null });
  const { total, next, operation } = state;

  const eventOperations = (button) => {
    const {
      target: {
        textContent: buttonName,
      },
    } = button;
    setState((state) => (
      calculate(state, buttonName)
    ));
  };

  return (
    <div className="calculator">
      <span className="display">
        { (next === null && total === null) ? 0 : `${total || ''} ${operation || ''} ${next || ''}` }
      </span>
      <Numbers eventOperations={eventOperations} />
      <Operators eventOperations={eventOperations} />
    </div>
  );
};

export default Calculator;
