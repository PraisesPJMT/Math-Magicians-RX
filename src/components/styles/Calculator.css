/* Calculator Styles */

.calculator {
  width: 50%;
  min-width: 350px;
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr;
  gap: 0.3rem;
  padding: 0.3rem;
  border: 1px solid #bcfd4c;
  box-shadow: 0 1px 10px #bcfd4c;
}

.calculator > span,
.calculator > button {
  padding: 1rem;
  border: none;
  font-size: 1.2rem;
  font-weight: 500;
}

.calculator > button:hover {
  font-weight: 900;
}

.number,
.operator {
  text-align: center;
  cursor: pointer;
}

.display {
  background: #6e6e6e99;
  grid-column: 1 / 5;
  grid-row: 1;
  text-align: right;
  color: #fff;
}

.orange {
  background: #bcfd4c;
  border-radius: 10px;
}

.light {
  background: #e0e0e0ff;
  border-radius: 10px;
}

/* Numbers */

.number-0 {
  grid-column: 1 / 3;
  grid-row: 6;
}

.number-1 {
  grid-column: 1;
  grid-row: 5;
}

.number-2 {
  grid-column: 2;
  grid-row: 5;
}

.number-3 {
  grid-column: 3;
  grid-row: 5;
}

.number-4 {
  grid-column: 1;
  grid-row: 4;
}

.number-5 {
  grid-column: 2;
  grid-row: 4;
}

.number-6 {
  grid-column: 3;
  grid-row: 4;
}

.number-7 {
  grid-column: 1;
  grid-row: 3;
}

.number-8 {
  grid-column: 2;
  grid-row: 3;
}

.number-9 {
  grid-column: 3;
  grid-row: 3;
}

/* Operators Styles */

.divide {
  grid-column: 4;
  grid-row: 2;
}

.times {
  grid-column: 4;
  grid-row: 3;
}

.minus {
  grid-column: 4;
  grid-row: 4;
}

.plus {
  grid-column: 4;
  grid-row: 5;
}

.equal {
  grid-column: 4;
  grid-row: 6;
}

/* Special Operators */

.clear {
  grid-column: 1;
  grid-row: 2;
}

.plus-minus {
  grid-column: 2;
  grid-row: 2;
}

.modulus {
  grid-column: 3;
  grid-row: 2;
}

.dot {
  grid-column: 3;
  grid-row: 6;
}

@media (max-width: 480px) {
  .calculator {
    min-width: 250px;
  }

  .calculator > span,
  .calculator > button {
    padding: 0.6rem;
    font-size: 1rem;
    font-weight: 500;
  }
}
