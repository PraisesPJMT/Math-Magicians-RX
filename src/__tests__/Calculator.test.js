import { render, screen, fireEvent } from '@testing-library/react';
import Calculator from '../components/Calculator';

require('@testing-library/jest-dom');

describe('Tests for Calculator component', () => {
  test('Renders the Calculator component', () => {
    const calculator = render(<Calculator />);
    expect(calculator).toMatchSnapshot();
  });

  test('Simulate user pressing numbers to display right numbers', () => {
    render(<Calculator />);
    const calculatorDisplay = document.querySelector('.display');
    const number4 = screen.getByRole('button', { name: '4' });
    const number0 = screen.getByRole('button', { name: '0' });
    const number1 = screen.getByRole('button', { name: '1' });
    fireEvent.click(number4);
    fireEvent.click(number0);
    fireEvent.click(number1);
    fireEvent.click(number0);
    expect(calculatorDisplay.textContent).toBe('  4010');
  });

  /* User performing pressing operations */

  test('Simulate user pressing the plus operator to display \'+\'', () => {
    render(<Calculator />);
    const calculatorDisplay = document.querySelector('.display');
    const plusOp = screen.getByRole('button', { name: '+' });
    fireEvent.click(plusOp);
    expect(calculatorDisplay.textContent).toBe(' + ');
  });

  test('Simulate user pressing the minus operator to display \'-\'', () => {
    render(<Calculator />);
    const calculatorDisplay = document.querySelector('.display');
    const plusOp = screen.getByRole('button', { name: '-' });
    fireEvent.click(plusOp);
    expect(calculatorDisplay.textContent).toBe(' - ');
  });

  test('Simulate user pressing the multiply operator to display \'×\'', () => {
    render(<Calculator />);
    const calculatorDisplay = document.querySelector('.display');
    const plusOp = screen.getByRole('button', { name: '×' });
    fireEvent.click(plusOp);
    expect(calculatorDisplay.textContent).toBe(' × ');
  });

  test('Simulate user pressing the divide operator to display \'÷\'', () => {
    render(<Calculator />);
    const calculatorDisplay = document.querySelector('.display');
    const plusOp = screen.getByRole('button', { name: '÷' });
    fireEvent.click(plusOp);
    expect(calculatorDisplay.textContent).toBe(' ÷ ');
  });

  test('Simulate user pressing the modulus operator to display \'%\'', () => {
    render(<Calculator />);
    const calculatorDisplay = document.querySelector('.display');
    const plusOp = screen.getByRole('button', { name: '%' });
    fireEvent.click(plusOp);
    expect(calculatorDisplay.textContent).toBe(' % ');
  });

  /* User performing valid operations */

  test('Simulate user performing addition (2 + 7)', () => {
    render(<Calculator />);
    const calculatorDisplay = document.querySelector('.display');
    const plusOp = screen.getByRole('button', { name: '+' });
    const number2 = screen.getByRole('button', { name: '2' });
    const number7 = screen.getByRole('button', { name: '7' });
    const equal = screen.getByRole('button', { name: '=' });
    fireEvent.click(number2);
    fireEvent.click(plusOp);
    fireEvent.click(number7);
    fireEvent.click(equal);
    expect(calculatorDisplay.textContent).toBe('9  ');
  });

  test('Simulate user performing subtraction (2 - 7)', () => {
    render(<Calculator />);
    const calculatorDisplay = document.querySelector('.display');
    const minusOp = screen.getByRole('button', { name: '-' });
    const number2 = screen.getByRole('button', { name: '2' });
    const number7 = screen.getByRole('button', { name: '7' });
    const equal = screen.getByRole('button', { name: '=' });
    fireEvent.click(number2);
    fireEvent.click(minusOp);
    fireEvent.click(number7);
    fireEvent.click(equal);
    expect(calculatorDisplay.textContent).toBe('-5  ');
  });

  test('Simulate user performing multiplication (2 × 7)', () => {
    render(<Calculator />);
    const calculatorDisplay = document.querySelector('.display');
    const multOp = screen.getByRole('button', { name: '×' });
    const number2 = screen.getByRole('button', { name: '2' });
    const number7 = screen.getByRole('button', { name: '7' });
    const equal = screen.getByRole('button', { name: '=' });
    fireEvent.click(number2);
    fireEvent.click(multOp);
    fireEvent.click(number7);
    fireEvent.click(equal);
    expect(calculatorDisplay.textContent).toBe('14  ');
  });

  test('Simulate user performing division (14 ÷ 7)', () => {
    render(<Calculator />);
    const calculatorDisplay = document.querySelector('.display');
    const divOp = screen.getByRole('button', { name: '÷' });
    const number1 = screen.getByRole('button', { name: '1' });
    const number4 = screen.getByRole('button', { name: '4' });
    const number7 = screen.getByRole('button', { name: '7' });
    const equal = screen.getByRole('button', { name: '=' });
    fireEvent.click(number1);
    fireEvent.click(number4);
    fireEvent.click(divOp);
    fireEvent.click(number7);
    fireEvent.click(equal);
    expect(calculatorDisplay.textContent).toBe('2  ');
  });

  test('Simulate user performing modulus (7 % 2)', () => {
    render(<Calculator />);
    const calculatorDisplay = document.querySelector('.display');
    const modOp = screen.getByRole('button', { name: '%' });
    const number2 = screen.getByRole('button', { name: '2' });
    const number7 = screen.getByRole('button', { name: '7' });
    const equal = screen.getByRole('button', { name: '=' });
    fireEvent.click(number7);
    fireEvent.click(modOp);
    fireEvent.click(number2);
    fireEvent.click(equal);
    expect(calculatorDisplay.textContent).toBe('1  ');
  });

  /* User performing invalid operations */

  test('Simulate user performing invalid division (14 ÷ 0)', () => {
    render(<Calculator />);
    const calculatorDisplay = document.querySelector('.display');
    const divOp = screen.getByRole('button', { name: '÷' });
    const number1 = screen.getByRole('button', { name: '1' });
    const number4 = screen.getByRole('button', { name: '4' });
    const number0 = screen.getByRole('button', { name: '0' });
    const equal = screen.getByRole('button', { name: '=' });
    fireEvent.click(number1);
    fireEvent.click(number4);
    fireEvent.click(divOp);
    fireEvent.click(number0);
    fireEvent.click(equal);
    expect(calculatorDisplay.textContent).toBe('Can\'t divide by 0.  ');
  });

  test('Simulate user performing invalid modulus (7 % 0)', () => {
    render(<Calculator />);
    const calculatorDisplay = document.querySelector('.display');
    const modOp = screen.getByRole('button', { name: '%' });
    const number0 = screen.getByRole('button', { name: '0' });
    const number7 = screen.getByRole('button', { name: '7' });
    const equal = screen.getByRole('button', { name: '=' });
    fireEvent.click(number7);
    fireEvent.click(modOp);
    fireEvent.click(number0);
    fireEvent.click(equal);
    expect(calculatorDisplay.textContent).toBe('Can\'t find modulo as can\'t divide by 0.  ');
  });
});
