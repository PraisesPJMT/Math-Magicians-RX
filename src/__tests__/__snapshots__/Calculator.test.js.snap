// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for Calculator component Renders the Calculator component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="calculator"
      >
        <span
          class="display"
        >
          0
        </span>
        <button
          class="number light number-0"
          type="button"
        >
          0
        </button>
        <button
          class="number light number-1"
          type="button"
        >
          1
        </button>
        <button
          class="number light number-2"
          type="button"
        >
          2
        </button>
        <button
          class="number light number-3"
          type="button"
        >
          3
        </button>
        <button
          class="number light number-4"
          type="button"
        >
          4
        </button>
        <button
          class="number light number-5"
          type="button"
        >
          5
        </button>
        <button
          class="number light number-6"
          type="button"
        >
          6
        </button>
        <button
          class="number light number-7"
          type="button"
        >
          7
        </button>
        <button
          class="number light number-8"
          type="button"
        >
          8
        </button>
        <button
          class="number light number-9"
          type="button"
        >
          9
        </button>
        <button
          class="operator orange plus"
          type="button"
        >
          +
        </button>
        <button
          class="operator orange minus"
          type="button"
        >
          -
        </button>
        <button
          class="operator orange divide"
          type="button"
        >
          ÷
        </button>
        <button
          class="operator orange times"
          type="button"
        >
          ×
        </button>
        <button
          class="operator orange equal"
          type="button"
        >
          =
        </button>
        <button
          class="operator light modulus"
          type="button"
        >
          %
        </button>
        <button
          class="operator light plus-minus"
          type="button"
        >
          +/-
        </button>
        <button
          class="operator light dot"
          type="button"
        >
          .
        </button>
        <button
          class="operator light clear"
          type="button"
        >
          AC
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="calculator"
    >
      <span
        class="display"
      >
        0
      </span>
      <button
        class="number light number-0"
        type="button"
      >
        0
      </button>
      <button
        class="number light number-1"
        type="button"
      >
        1
      </button>
      <button
        class="number light number-2"
        type="button"
      >
        2
      </button>
      <button
        class="number light number-3"
        type="button"
      >
        3
      </button>
      <button
        class="number light number-4"
        type="button"
      >
        4
      </button>
      <button
        class="number light number-5"
        type="button"
      >
        5
      </button>
      <button
        class="number light number-6"
        type="button"
      >
        6
      </button>
      <button
        class="number light number-7"
        type="button"
      >
        7
      </button>
      <button
        class="number light number-8"
        type="button"
      >
        8
      </button>
      <button
        class="number light number-9"
        type="button"
      >
        9
      </button>
      <button
        class="operator orange plus"
        type="button"
      >
        +
      </button>
      <button
        class="operator orange minus"
        type="button"
      >
        -
      </button>
      <button
        class="operator orange divide"
        type="button"
      >
        ÷
      </button>
      <button
        class="operator orange times"
        type="button"
      >
        ×
      </button>
      <button
        class="operator orange equal"
        type="button"
      >
        =
      </button>
      <button
        class="operator light modulus"
        type="button"
      >
        %
      </button>
      <button
        class="operator light plus-minus"
        type="button"
      >
        +/-
      </button>
      <button
        class="operator light dot"
        type="button"
      >
        .
      </button>
      <button
        class="operator light clear"
        type="button"
      >
        AC
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
